### DISCUSSION TITLE
Question about nested objects and .save_related()

### DISCUSSION THREAD

**Discussion by franzwollang** [collerek/ormar#314] [Q&A] [answered]
I'm converting over from using SQLAlchemy ORM and I'm having some trouble with getting nested objects to save correctly.

I have models and relationships set up like so:
   Vendor <-- Restaurant <-- Menu <-- MenuSection <-- MenuItem <-- Offer
   Vendor <--> Employee
   Order --> Vendor
   
In order to seed the database for testings purposes, I'm adding some new employees, a vendor, a restaurant, and some orders. I create the vendor object w/ nested objects in it, and the order object. Then I run the following code:

```
await vendor.save()

        for employee in vendor.employees: # adding employees into DB manually because M2M relationship 
                await employee.save()

await vendor.save_related(follow=True, save_all=True)
await order.save()
```

After executing this code, the employees, vendor, and order get added correctly (including relationships) but the restaurants, menus, etc... don't get added. If I manually go in and add a loop to call .save() on each of the restaurant objects first, then they will also get added correctly (including relationships) but then menus, menuSections, etc... don't get added. If I do the same manually for menus, then it works but menuSections, menuItems, etc... don't get added.

I feel like I'm missing something important about how .save_related() is supposed to work. Any ideas? I'm getting pretty frustrated trying to figure this out. Help would be much appreciated.

**Comment by collerek** [comment, marked as answer]
You don't provide a full code sample but I would assume you provide a primary keys for your deeply nested relations. 

Save related is a shortcut function that executes upsert on each model. The insert/update operation is based on the presence of the primary keys. If the primary key is present ormar assumes it's an update (otherwise you would get an integrity error with duplicated primary key if in fact the given models already exists and you try to insert them again). 

So if you want to create the full tree at once (so all models are not existing) the child models cannot have primary keys set (so should be autoincrement or columns with default). I can add something like force_insert flag but then it's on you to handle the integrity errors.

I want to avoid running a separate query to check if the model already exists for each instance as that would slow down the whole process substantiality.

**Reply by franzwollang** [reply]
I see. Thank you ten times over for your help, so quickly, and even on a Sunday evening! You are a programming saint. I briefly thought about the primary keys being set and if it would be a problem, but I didn't think about the concerns you noted. I will fix my problem by leaving the primary keys nulled, as you suggested.

I think a flag to force insert would be nice; I'm a fan of keeping the default behavior clean and safe, but giving people the option to shoot themselves in the foot if they want to. The extra flexibility can often make it easier to do specific things, as long as you know what you're doing.


While we're on the topic of nested models, there is a note in the documentation stating the inserts are done in many different SQL calls. What are the obstacles to making it possible to insert nested models in fewer (or even a single) SQL call? I would be willing to help work on this over the next few months if there is a possible path towards implementing this kind of optimization.

**Reply by franzwollang** [reply]
@collerek I removed the primary keys like you suggested, but it is still failing. I'm using non-integer primary keys with a default; upon inspection, it looks like the 'default' is applied to the empty primary key fields when the entities are instantiated, so when I then call .save_related() it ends up being the same as if I had just manually inserted the keys.

Is this an Ormar bug or am I still doing something wrong?

**Reply by collerek** [reply]
Yes, to keep it empty you would have to use `server_default`, pydantic is populating the default values on init.

That would require this `force_insert` flag if you know all of them are new. (you can open an issue for this).

BTW. **Thank you** for your sponsoring 😄

**Reply by franzwollang** [reply]
Ok, thank you for explaining.

My pleasure. Thank you for stepping up to create this project; there is no doubt that there is a lot of value in not having to maintain two models for every object when an API starts to get large, plus this further promotes higher cohesion by incentivizing one to write import/export processing code also in the single Ormar model using Pydantic pre-validators and decoders rather than as a separate translation layer between Pydantic and SQLAlchemy ORM models. I also like the way you are approaching the querying features in contrast to the SQLAlchemy ORM, which feels more rigid and less transparent.

**Comment by collerek** [comment]
Thanks for the nice words😁

I assume you mean multiple related models of the same type/class are inserted separately, why not issue one query per model? The answer is mainly the complexity and primary keys. 

Since depending on the side of the relation and which one is which (fk or m2m) you need to save sometimes the left side first, other times the right side first and obtain the primary keys to populate the relations. So it's never a linear save `a-b-c-d` (well it can be, so never say never but that's just one of many cases). Very well you might have to save a than d than c and b at the end. 

So in order to save all models, keep the corresponding primary keys after save (bulk save return none in SQL so you need additional query there to fetch inserted primary keys and server defaults), save proper models in order including through models for m2m relations it becomes much more complicated than you would think in a fists place. 

All in all you need to create something similar to prefetch query. (Which on master is a little nightmare, it's completely rewritten in composite primary keys branch 😅).

There might be also problems with primary key fetching after bulk save, since you can have simultaneous connections, so you cannot just return last inserted n rows as someone might have added something in the meantime, so you can break relations if you are not carefull.

**Reply by franzwollang** [reply]
I see.

My understanding is that a lot of the complexity is coming from the fact that you have to wait for the SQL engine to assign the keys; for auto-incremented integer keys, there isn't really any way around the problem since a separate call could insert something else while your call is doing things and so you can't make assumptions about what the assigned keys will be.

For UUIDs, however, one can rely on the uniqueness of the key even if it is generated first and sent to the SQL engine; integrity constraints enforced by the engine are then basically just a failsafe mechanism to prevent user error (accidentally using the same UUID twice) or weird replay errors (the same SQL query getting sent twice by accident). The ORM can then just manage these errors sensibly but otherwise operate under the assumption that it can generate and assign its own UUID primary keys.

Working on it from this angle, you could perform all the necessary calculations before sending a SQL query to figure out which tables need FK references to which other tables, both generate the UUID primary keys for the rows as necessary and add instructions to insert them into the appropriate tables (either directly for one-to-many or in through table for M2M), and then issue a single giant SQL statement.

Send this giant statement by itself if you don't care about the (extremely unlikely case) that there is an integrity error, or wrap it in a transaction w/ auto rollback and if there is an error automatically retry once with fresh UUIDs (solving the 99.999% of situations where the 0.0001% probability integrity error actually occurs).

What do you think? Is this something we could add as a special feature of using UUIDs instead of auto-incremented integers as PKs? The only benefit of integer PKs is smaller storage/memory requirements but for the vast majority of modern applications, I would assume it's much nicer to get faster write speed and have nearly guaranteed globally unique row identifiers (allowing you to be more flexible if you need to flatten the data or otherwise dissociate it from the original graph structure without having to do additional processing first) than to save some single digit percentage in storage/memory space.

While we're on this topic, how do you feel about supporting SLUG as an Ormar field type? It would be basically the same as the UUID field type, but instead of 32 character hex it would be 22 character URL-safe base64. This reduces the additional storage overhead concern of using UUIDs (if only slightly), and makes the IDs shorter (but still keeps it contained in potentially one byte per character if using UTF8) and more suitable for using in URLs and/or query params. It is my preferred form of UUID4.

**Reply by collerek** [reply]
That would work also for int primary keys if you provide them instead of relying on autoincrement.

One of the disadvantages of UUID (apart from size) is order :) I like to know which row was created first without the need of additional column (like created_date), and since UUID is "random" you cannot reliably order by this field.

As for the insert, how do you plan to insert multiple rows of data into multiple tables with one statement? That's not possible as far as I remember (there is an output function in tsql but that won't work for more than 2 tables I think - not to mention it's dialect-specific). You can have one statement by table (insert into with values list), but not one insert into several tables.

Uuid is already url-safe, so the benefit of encoding it to base64 would be only a smaller space used, but at the performance cost of encoding/decoding on the fly (unless you want to encode it and leave it like this).

**Reply by franzwollang** [reply]
It would work for int primary keys if you provide them but it would not work well with concurrency, since you would have to coordinate across async calls/threads/servers to decide who gets to use which integer. I know concurrency is not the highest priority for most users but it is good to account for if possible.

While you can order by an integer primary key, it is only a relative order not an absolute order; that may be enough for some typical business logic (like give me all customer orders with the newest first/last) but then often times you'll also want to know the date too (when was the order exactly?) for both pragmatic reasons and for record keeping for legal purposes and/or analytics purposes (time series prediction, anomaly prediction). Same logic applies to know exactly when many other specific actions/changes were made (e.g. maybe knowing a user account was made after all the other accounts doesn't help if I need to know the exact datetime, which might span a period of days or weeks, so I can cross-reference it with network logs to see if there was some funny business). Adding a datetime column doesn't seem like that much of a hassle given the benefits???

It doesn't have to be "one statement" per se (sorry for using imprecise language); what I meant was that it would be done "all at once" in a single "trip", or basically just avoiding multiple round trips of communication. All that would be required is the ability for both the client and SQL engine to handle multiple concurrent write requests. Even if the SQL engine will just handle them sequentially as they arrive, at least they all got sent together and generally the write latency is network bound not processor-bound. So it would basically be essentially the same as a single request to the engine since there's no round trip, and using a transaction is basically already an abstraction of allowing many actions to behave as an aggregate that singularly succeeds or fails.

A hex UUID is url-safe, but base64 isn't url-safe. I meant to specify using a URL-safe version of base64 encoding. I personally would leave it in base64 format. I don't think there is any downside to just using it as-is like that.

**Reply by collerek** [reply]
I meant providing integer values upfront - not providing one default function that serves all primary keys (like you can do with UUID as they don't rely on each other so you can use the same default generator or a different one - the probability of collision is extremely low anyway).

Yes, I fully agree that for production systems you need different requirements (audit etc.), but for quick tests, I prefer integers as I can tell if the expected outcome equals the actual one (i.e. ormar by default orders by pk - as in ormar query is always ordered).

**Reply by franzwollang** [reply]
So in conclusion, what does this mean for the possibility of 1) trying to speed up writes of data spanning many relationships by take advantage of generating IDs instead of letting the SQL engine assign them and 2) maybe supporting a SLUG type similar to the existing UUID type?

Are you in favor of exploring these changes or against them?

**Reply by collerek** [reply]
As for the data insert optimization -> sure - it can be done along with `force_insert` flag, I want this change to be nonbreaking.

As for the slug you can already save bytes field type as base64 encoded string, wouldn't that be enough? I assume the only difference is with the generation of the shorter encoded UUID into basae64, but since fields do not provide any default values - the user has to do this - what would be the logic of this field? Or if you prefer strings can't you provide a default generator that produces those base64 encoded strings for you and simply use string field?

**Reply by franzwollang** [reply]
Ok. Let me know if/how I can help. I'm not particularly well-versed with the SQLAlchemy core library, but I'm open helping if I can.

Yes, I can do that. Everything's good then.

**Reply by franzwollang** [reply]
On second thought, I realize that foreign key columns have an integrity constraint requiring the key they point to to already exist. So at minimum, you'd always have to set any row inserts first and only then could you update foreign key columns on models or in through tables. So there's always a minimum requirement of 2 batches of statements.

But if that's the case, then it doesn't matter whether you provide the key manually or let a default or auto-increment handle it. You have to wait for the row inserts anyway, so you could also just get the inserted keys as well before doing the foreign key column updates.

So I guess the best that could be done is ensure that for nested objects across related tables, inserts always happen in exactly two batches of statements. So the latency would be around two round trips worth. Don't know if that's how it already works; in any case, sorry for not thinking it through carefully before discussing it with you.

**Reply by collerek** [reply]
If all keys on both sides will have keys populated you can insert them one after another in one transaction. 
Only then you can issue queries in batch - if you have autoincrement keys there is no easy way to populate and retrieve the values in batch mode to populate related values on different models.

Yeah, that's what I meant when I wrote that one time you need to save left side first and some times right time first (depending on fk -> if it's reversed or normal relation (a->b or b->a with a-b-c-d relation)).

Since right now there is upsert going it does not work in batch. The `force_insert` flag will work only when the primary keys will be already populated (so set by hand or default function provided), even with autoincrement but values not set it will raise an exception as this will not work in batch mode.
